src/lib/utils/DY.Lib.Printing.fst

```diff
@@ -50,8 +39,27 @@ let rec uint_list_to_string seq =
     else Printf.sprintf "%c%s" (FStar.Char.char_of_int (FStar.UInt8.v hd)) (uint_list_to_string tl)
   )
 
+// Allow the warning "Global binding 'DY.Lib.Printing.usage_to_string' is recursive but not used in its body",
+// because although `usage` and `bytes` are mutually recursive,
+// in practice the function `bytes_to_string` do not call the function `usage_to_string`.
+
+#push-options "--warn_error '+328'"
+val usage_to_string: (u:usage) -> string
 val bytes_to_string: (b:bytes) -> string
-let rec bytes_to_string b =
+let rec usage_to_string u =
+  match u with
+  | NoUsage -> "NoUsage"
+  | SigKey tag -> "SigKey " ^ tag
+  | SigNonce -> "SigNonce"
+  | PkdecKey tag -> "PkdecKey " ^ tag
+  | PkNonce -> "PkNonce"
+  | AeadKey tag -> "AeadKey " ^ tag
+  | DhKey tag -> "DhKey " ^ tag
+  | KdfExtractSaltKey tag data -> Printf.sprintf "KdfExtractSaltKey %s (data=(%s))" tag (bytes_to_string data)
+  | KdfExtractIkmKey tag data -> Printf.sprintf "KdfExtractIkmKey %s (data=(%s))" tag (bytes_to_string data)
+  | KdfExpandKey tag data -> Printf.sprintf "KdfExpandKey %s (data=(%s))" tag (bytes_to_string data)
+
+and bytes_to_string b =
```

Comment:
It is not needed because we don't print the usage in the `Rand` constructor, so using the `let rec ... and ...` is more future-proof if someday we want to print the usage. But if you prefer for now we can define `bytes_to_string` before.

Reply:
I think it would be nicer to keep it separate for now but you can also keep it like that.

Reply:
I kept them separate then!